src/app/src/main/kotlin/org/imperial/mrc/hint/logic/UserLogic.kt

```diff
@@ -0,0 +1,87 @@
+package org.imperial.mrc.hint.logic
+
+import org.imperial.mrc.hint.db.UserRepository
+import org.imperial.mrc.hint.emails.EmailManager
+import org.imperial.mrc.hint.emails.PasswordEmailTemplate
+import org.imperial.mrc.hint.exceptions.UserException
+import org.pac4j.core.profile.CommonProfile
+import org.pac4j.sql.profile.DbProfile
+import org.pac4j.sql.profile.service.DbProfileService
+import org.springframework.stereotype.Component
+import java.security.SecureRandom
+import java.util.*
+
+interface UserLogic {
+    fun addUser(email: String, password: String?)
+    fun removeUser(email: String)
+    fun getUser(email: String): CommonProfile?
+    fun updateUserPassword(user: CommonProfile, password: String)
+}
+
+@Component
+class DbProfileServiceUserLogic(private val userRepository: UserRepository,
+                                private val profileService: DbProfileService,
+                                private val emailManager: EmailManager) : UserLogic {
+
+    override fun addUser(email: String, password: String?) {
+
+        if (!email.contains("@")) {
+            throw UserException("Please provide a valid email address")
+        }
+
+        if (getUser(email) != null) {
+            throw UserException("User already exists")
+        }
+
+        val profile = DbProfile()
+        profile.build(email, mapOf("username" to email))
+
+        val pw = if (password.isNullOrEmpty()) {
+            val pw = generateRandomPassword()
+            emailManager.sendPasswordEmail(email, email, PasswordEmailTemplate.CreateAccount())
+            pw
+        } else {
+            password
+        }
+        profileService.create(profile, pw)
+    }
+
+    override fun removeUser(email: String) {
+        val user = getUser(email) ?: throw UserException("User does not exist")
+        profileService.removeById(user.id)
+    }
+
+    override fun getUser(email: String): CommonProfile? {
+
+        val emailArray = email.split("@")
+        if (emailArray.count() == 1) {
+            throw UserException("Please provide a valid email address")
+        }
+
+        val caseInsensitiveDomainRegex = Regex("(?-i)${emailArray[0]}@(?i)${emailArray[1]}")
```

Comment:
This means we'll stay case sensitive for the name part of the email address - is that what we want?

Reply:
Yep, emails are case sensitive in general.

Reply:
Well, in theory they could be case sensitive. In practice, unlikely. https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive

Reply:
Yeah you're right, they're meant to be case-sensitive, but most providers let you get away with variants. So users might be used to treating them as case insensitive and being a bit cavalier with using variants. I think it's fine to keep it case sensitive for our purposes and see if anyone objects.